### PR TITLE
Add a deploy command

### PR DESCRIPTION
**Pull request by ironcladlou**
Add a deploy command which allows users to view, start, and restart
deployments manually.

### CODE REVIEW

**Comment by smarterclayton** [comment]
cc @fabianofranz

----- Original Message -----

**Comment by fabianofranz** [comment]
Very nice @ironcladlou.

I'll work on the command for listing the deployment history then, @smarterclayton should history be a separate command like `osc deployments` or something around this one, like `osc deploy -L`? I like the idea of a `-L` flag for listing, but we need to have in mind that this command will probably have additional flags for filtering and changing the level of details to be listed.

**Comment by ironcladlou** [comment]
@fabianofranz Yeah, I kind of like the idea of putting more stuff in this command as well, but I'm not sure how constrained in scope it should ultimately be. @smarterclayton, any thoughts here? I can see this command growing, but I wonder if there are any possible conflicts with other more "generic" commands like `describe dc` and `get rc`.

**Comment by pmorie** [comment]
Do describers take arguments to filter the info they display at all?

**Comment by fabianofranz** [comment]
BTW I was thinking about adding to Cobra some kind of flag grouping, probably through multiple  `FlagSet`, which could be a solution for this and other commands where we want to have multiple features in one command, and flags that only apply to certain of those features. For example, something like:

```
$ osc deploy -h
View, start and restart deployments and list deployments history.
(...)

Deploy options:
     --force       Start a new deployment now.
     --retry       Retry the latest failed deployment.

List deployments options (applies to -L):
  -L  --list       List the latest deployments in history.
      --limit=n    Limit the number of items to display in history to "n".
  -v  --verbose    Verbose output.
(...)
```

@smarterclayton thoughts?

**Comment by ironcladlou** [comment]
@smarterclayton @fabianofranz 

Do you think this is good enough for a minimal first pass? Just having force and retry is immediately useful. As we flesh out stopping deployments and handling concurrent deployment runs, we'll build on this command.

File: pkg/cmd/cli/cmd/deploy.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+		},
+	}
+
+	cmd.Flags().BoolVar(&forceDeploy, "force", false, "Start a new deployment now.")
+	cmd.Flags().BoolVar(&retryDeploy, "retry", false, "Retry the latest failed deployment.")
+
+	return cmd
+}
+
+func RunForceDeployment(config *deployapi.DeploymentConfig, kubeClient kclient.Interface, osClient client.Interface, out io.Writer) error {
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/deploy.go]: Do these need to be public?

Comment (ironcladlou) [inline pkg/cmd/cli/cmd/deploy.go]: Nope, just an oversight. Fixed.

File: pkg/cmd/cli/cmd/deploy.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+		},
+	}
+
+	cmd.Flags().BoolVar(&forceDeploy, "force", false, "Start a new deployment now.")
+	cmd.Flags().BoolVar(&retryDeploy, "retry", false, "Retry the latest failed deployment.")
+
+	return cmd
+}
+
+func runForceDeployment(config *deployapi.DeploymentConfig, kubeClient kclient.Interface, osClient client.Interface, out io.Writer) error {
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/deploy.go]: Add a command test that mocks some of it (see what I've done in https://github.com/smarterclayton/origin/commit/b27db760b83bd6466ce258f41f1c8049da51e169 for start-build).  Also add a hack/test-cmd.  Then you can be merged.

File: pkg/cmd/cli/cmd/deploy.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+				d := describe.NewLatestDeploymentDescriber(osClient, kubeClient)
+				s, err := d.Describe(config.Namespace, config.Name)
+				cmdutil.CheckErr(err)
+				fmt.Printf(s)
+			}
+			cmdutil.CheckErr(err)
+		},
+	}
+
+	cmd.Flags().BoolVar(&forceDeploy, "force", false, "Start a new deployment now.")
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/deploy.go]: Actually, let's make this "--latest" (or similar) so we can reserve "--force" for overriding problems.

Comment (ironcladlou) [inline pkg/cmd/cli/cmd/deploy.go]: Good idea, I like --latest.

**Comment by ironcladlou** [comment]
@smarterclayton Refactored, polished output, added lots of unit test coverage, added a little bit to test-cmd.sh. PTAL.

File: pkg/cmd/cli/cmd/deploy.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+	}
+	return err
+}
+
+func statusFor(deployment *kapi.ReplicationController) deployapi.DeploymentStatus {
+	return deployapi.DeploymentStatus(deployment.Annotations[deployapi.DeploymentStatusAnnotation])
+}
+
+// deployCommandClientImpl is a pluggable deployCommandClient.
+type deployCommandClientImpl struct {
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/deploy.go]: This abstraction is a bit much (you should just take DeploymentConfigNamespacer instead)

**Comment by smarterclayton** [comment]
One minor comment, LGTM otherwise

File: pkg/cmd/cli/cmd/deploy.go (hunk 4 above)
Comment (ironcladlou) [inline pkg/cmd/cli/cmd/deploy.go]: I can revisit the interface usage the next time we come in here to add more stuff.

**Comment by smarterclayton** [comment]
[merge]

**Comment by ironcladlou** [comment]
TestDNS flaked, [merge] again

**Comment by smarterclayton** [comment]
Did you hang?

**Pull request merged**